src/transformers/safetensors_conversion.py

```diff
@@ -0,0 +1,99 @@
+from typing import Optional
+
+from huggingface_hub import Discussion, HfApi
+
+from .utils import cached_file, logging
+
+
+logger = logging.get_logger(__name__)
+
+
+def previous_pr(api: "HfApi", model_id: str, pr_title: str) -> Optional["Discussion"]:
+    try:
+        main_commit = api.list_repo_commits(model_id)[0].commit_id
+        discussions = api.get_repo_discussions(repo_id=model_id)
+    except Exception:
+        return None
+    for discussion in discussions:
+        if discussion.status == "open" and discussion.is_pull_request and discussion.title == pr_title:
+            commits = api.list_repo_commits(model_id, revision=discussion.git_reference)
+
+            if main_commit == commits[1].commit_id:
+                return discussion
+    return None
+
+
+def spawn_conversion(token: str, model_id: str):
+    print("Sending conversion request")
+    import asyncio
+    import json
+    import uuid
+
+    import websockets
```

Comment:
This is unfortunately necessary I think.

Gradio will skip the queue if you use the REST API, leading to swarming the space very fast instead of waiting in queue.
Also if users cancel action, the request will stay in the queue regardless of the user presence in it. 
In the case of the websocket, if the websocket connection is severed, the space will remove it from the queue.


@abidlabs Could you confirm/infirm ? (The context is that we want to trigger a space action from this bit of code, ideally without websockets dependency, with all the properties I mention above which I think are exclusive to WS connections).

Reply:
FWIW Gradio has dropped websockets recently (in favor of SSE) if i'm not mistaken, cc @abidlabs 

I'm also in favor of dropping the requirement here and using SSE instead. (ping @Wauplin maybe)

Reply:
If the properties of queuing + dropping correctly are preserved I'm more than happy to drop !

Reply:
Yes, Gradio 4.0 or higher uses SSE. We also now run every task through the queue by default (aka the rest api is disabled by default), so this should not be necessary.

Reply:
Switch from WebSocket to SSE in https://github.com/huggingface/transformers/pull/27656